### PR TITLE
Add `defines` to vm::docs

### PR DESCRIPTION
**Pull request by kantai**
This PR:

* Fixes `eq?` in the current docgen
* Adds `define`, `define-map`, `define-read-only`, `define-public` to `vm::docs::make_json_api_reference()`

Resolves Issue #1016 

You can test this either locally or with the built docker image for this branch. For example:

```
$ docker run -it quay.io/blockstack/blockstack-core:feature_add-defines-to-docs blockstack-core docgen
```

### CODE REVIEW

File: src/vm/docs/mod.rs [hunk 1]
```diff
@@ -205,6 +195,18 @@ fn make_for_simple_native(api: &SimpleFunctionAPI, function: &NativeFunctions) -
     }
 }
 
+const EQUALS_API: SpecialAPI = SpecialAPI {
+    name: "eq?",
+    input_type: "A, A, ...",
+    output_type: "bool",
+    signature: "(eq? v1 v2...)",
+    description: "Compares the inputted values, returning `true` if they are all equal. Note that _unlike_ the `(and ...)` function, `(eq? ...)` will _not_ short-circuit.",
```
Comment (jcnelson) [inline src/vm/docs/mod.rs]: Ah, good point.  Do we want `eq?` to short-circuit?

File: src/vm/docs/mod.rs [hunk 2]
```diff
@@ -538,6 +540,96 @@ The `header-hash`, `burnchain-header-hash`, and `vrf-seed` properties return a 3
...
+    description: "`define-public` is used to define a _public_ function and transaction for a smart contract. Public
+functions are callable from other smart contracts, and may be invoked directly by users by submitting a transaction
+to the Stacks blockchain.
+
+Like other kinds of definition statements, `define-public` may only be used at the top level of a smart contract
+definition (i.e., you cannot put a define statement in the middle of a function body).
+
+Public functions _must_ return a ResponseType (using either `ok` or `err`). Any datamap modifications performed by
+a public function will be aborted if the function returns an `err` type. Public functions may be invoked by other
+contracts via `contract-call!`.",
```
Comment (jcnelson) [inline src/vm/docs/mod.rs]: Even if the function invoking `contract-call!` handles the `(err` result, no changes will be materialized?

Comment (kantai) [reply src/vm/docs/mod.rs]: Correct -- anytime a function returns `err`, any of the modifications it (or functions it invoked) made are aborted. If you want a function to return different kinds of values while still making database changes, you would have to use the `ok` type.

Comment (kantai) [reply src/vm/docs/mod.rs]: The _caller_, if it handles the `err` result can still materialize its own changes, though.

File: src/vm/docs/mod.rs (hunk 1 above)
Comment (kantai) [reply src/vm/docs/mod.rs]: Hmm, it doesn't usually in Lisps, and I'm inclined to follow that convention.

Comment (jcnelson) [reply src/vm/docs/mod.rs]: Okay, sounds good.  Let's just follow the convention (but thanks for mentioning it explicitly here).

File: src/vm/docs/mod.rs (hunk 2 above)
Comment (jcnelson) [reply src/vm/docs/mod.rs]: That's what I thought.  Thanks for clarifying!

**Comment by moxiegirl** [comment]
@kantai I'm getting an error processing the JSON here: 

```
$ docker run -it -v $HOME/blockstack-dev-data:/data/ 057d094d8764 blockstack-core docgen | jsonpp > ~/repos/docs.blockstack/_data/clarityRef.json
ERROR: Broken json on line 1, char 2: unexpected end of JSON input
                                                                    Context: 
``

File: src/vm/docs/mod.rs [hunk 3]
```diff
@@ -538,6 +540,96 @@ The `header-hash`, `burnchain-header-hash`, and `vrf-seed` properties return a 3
...
+be invoked by other contracts via `contract-call!`.",
+    example: "
+(define-read-only (just-return-one-hundred) 
+  (* 10 10))"
+};
+
+const DEFINE_MAP_API: SpecialAPI = SpecialAPI {
+    name: "define-map",
+    input_type: "MapName, KeyTupleDefinition, MapTupleDefinition",
+    output_type: "",
```
Comment (moxiegirl) [inline src/vm/docs/mod.rs]: For `output_type: does `user defined function` or "Not applicable" make sense?

Comment (kantai) [reply src/vm/docs/mod.rs]: yeah, that would work

Comment (kantai) [reply src/vm/docs/mod.rs]: switched to "Not Applicable"

**Pull request merged**